zkRegistry.Destroy(): negative Waitgroup counter

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

-->


**What happened**:

When I try to invoke the `registryProtocol.Destroy()` to destroy all the registry instances, I got the error:
![image](https://user-images.githubusercontent.com/9923838/68590629-03178780-04ca-11ea-9d3b-548eaa6bb58a.png)

I div into the source code and found that the `registryProtocol.Destroy()` will invoke the `zkRegistry.Destroy()`:

![image](https://user-images.githubusercontent.com/9923838/68590804-5be72000-04ca-11ea-8f00-079d9b5144f3.png)

But the Destroy() will issue the callback:
![image](https://user-images.githubusercontent.com/9923838/68591382-97ceb500-04cb-11ea-9d31-7889d88a371f.png)
The line which has ba breakpoint will be invoked then crash. The `listener.Next()` will step into:

![image](https://user-images.githubusercontent.com/9923838/68591579-01e75a00-04cc-11ea-9627-4aa1da2c13a4.png)

I doubt that the method `RegistryConfigurationListener.Close()` was invoked multiple times.

Here is the log from my computer ( some log was added by myself to debug, just ignore them):
```
2019-11-10T22:08:41.199+0800    INFO    protocol/protocol.go:298        registryProtocol destroy.
2019-11-10T22:08:41.199+0800    INFO    protocol/protocol.go:126        Exporter unexport.
2019-11-10T22:08:41.199+0800    INFO    protocol/invoker.go:67  Destroy invoker: dubbo://:20000/UserProvider?accesslog=&app.version=0.0.1&application=BDTService&bean.name=UserProvider&cluster=failover&environment=dev&execute.limit=3&execute.limit.rejected.handler=&group=&interface=com.ikurento.user.UserProvider&loadbalance=random&methods.GetUser.loadbalance=random&methods.GetUser.retries=1&methods.GetUser.tps.limit.interval=&methods.GetUser.tps.limit.rate=&methods.GetUser.tps.limit.strategy=&methods.GetUser.weight=0&module=dubbogo+user-info+server&name=BDTService&organization=ikurento.com&owner=ZX&registry.role=3&retries=&service.filter=echo%2Ctoken%2Caccesslog%2Ctps%2Cexecute%2Cactive&timestamp=1573394909&token=3b08d342-c576-4b72-85f6-855ddacdd502&tps.limit.interval=&tps.limit.rate=&tps.limit.rejected.handler=&tps.limit.strategy=&tps.limiter=&version=&warmup=100
2019-11-10T22:08:41.199+0800    INFO    protocol/protocol.go:308        Exporter unexpor: dubbo://:20000/UserProvider?accesslog=&app.version=0.0.1&application=BDTService&bean.name=UserProvider&cluster=failover&environment=dev&execute.limit=3&execute.limit.rejected.handler=&group=&interface=com.ikurento.user.UserProvider&loadbalance=random&methods.GetUser.loadbalance=random&methods.GetUser.retries=1&methods.GetUser.tps.limit.interval=&methods.GetUser.tps.limit.rate=&methods.GetUser.tps.limit.strategy=&methods.GetUser.weight=0&module=dubbogo+user-info+server&name=BDTService&organization=ikurento.com&owner=ZX&registry.role=3&retries=&service.filter=echo%2Ctoken%2Caccesslog%2Ctps%2Cexecute%2Cactive&timestamp=1573394909&token=3b08d342-c576-4b72-85f6-855ddacdd502&tps.limit.interval=&tps.limit.rate=&tps.limit.rejected.handler=&tps.limit.strategy=&tps.limiter=&version=&warmup=100
2019-11-10T22:08:41.199+0800    INFO    protocol/protocol.go:315        Destroy registry: zookeeper://:@127.0.0.1:2181/?interface=&group=&version=
2019-11-10T22:08:41.199+0800    WARN    zookeeper/facade.go:55  (ZkProviderRegistry)reconnectZkRegistry goroutine exit now...
github.com/apache/dubbo-go/common/logger.Warnf
        /Users/mindeng/git/apache-dubbo/dubbo-go/common/logger/logging.go:36
github.com/apache/dubbo-go/remoting/zookeeper.HandleClientRestart
        /Users/mindeng/git/apache-dubbo/dubbo-go/remoting/zookeeper/facade.go:55
2019-11-10T22:08:41.199+0800    WARN    zookeeper/listener.go:95        zk consumer register has quit, so zk event listener exit now.
github.com/apache/dubbo-go/common/logger.Warnf
        /Users/mindeng/git/apache-dubbo/dubbo-go/common/logger/logging.go:36
github.com/apache/dubbo-go/registry/zookeeper.(*RegistryConfigurationListener).Next
        /Users/mindeng/git/apache-dubbo/dubbo-go/registry/zookeeper/listener.go:95
github.com/apache/dubbo-go/registry/zookeeper.(*zkRegistry).Subscribe
        /Users/mindeng/git/apache-dubbo/dubbo-go/registry/zookeeper/registry.go:423
2019-11-10T22:08:41.199+0800    INFO    zookeeper/registry.go:473       begin to close provider zk client
2019-11-10T22:08:41.199+0800    WARN    zookeeper/registry.go:424       Selector.watch() = error{listener stopped}
github.com/apache/dubbo-go/common/logger.Warnf
        /Users/mindeng/git/apache-dubbo/dubbo-go/common/logger/logging.go:36
github.com/apache/dubbo-go/registry/zookeeper.(*zkRegistry).Subscribe
        /Users/mindeng/git/apache-dubbo/dubbo-go/registry/zookeeper/registry.go:424
panic: sync: negative WaitGroup counter

goroutine 197 [running]:
sync.(*WaitGroup).Add(0xc000324020, 0xffffffffffffffff)
        /usr/local/go/src/sync/waitgroup.go:74 +0x135
sync.(*WaitGroup).Done(...)
        /usr/local/go/src/sync/waitgroup.go:99
github.com/apache/dubbo-go/registry/zookeeper.(*RegistryConfigurationListener).Close(0xc000308480)
        /Users/mindeng/git/apache-dubbo/dubbo-go/registry/zookeeper/listener.go:112 +0x3f
github.com/apache/dubbo-go/registry/zookeeper.(*zkRegistry).Subscribe(0xc000324000, 0xc0002fa410, 0x1637ec0, 0xc0002f0720)
        /Users/mindeng/git/apache-dubbo/dubbo-go/registry/zookeeper/registry.go:425 +0x52a
created by github.com/apache/dubbo-go/registry/protocol.(*registryProtocol).Export
        /Users/mindeng/git/apache-dubbo/dubbo-go/registry/protocol/protocol.go:177 +0x85f

Process finished with exit code 2
```
**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:
I reproduce this problem by using the `dubbogo-samples` go server, and the code contains this [Fix: The registryProtocol should be singleton](https://github.com/apache/dubbo-go/pull/251)

![image](https://user-images.githubusercontent.com/9923838/68592053-2263e400-04cd-11ea-894a-bec621411e90.png)

**Anything else we need to know?**:

--- a/registry/protocol/protocol.go
+++ b/registry/protocol/protocol.go
@@ -338,10 +338,10 @@ func setProviderUrl(regURL *common.URL, providerURL *common.URL) {
 }
 
 func GetProtocol() protocol.Protocol {
-	if regProtocol != nil {
-		return regProtocol
+	if regProtocol == nil {
+		regProtocol = newRegistryProtocol()
 	}
-	return newRegistryProtocol()
+	return regProtocol
 }
 
 type wrappedInvoker struct {

--- a/registry/protocol/protocol_test.go
+++ b/registry/protocol/protocol_test.go
@@ -291,3 +291,8 @@ func TestExportWithApplicationConfig(t *testing.T) {
 	v2, _ := regProtocol.bounds.Load(getCacheKey(newUrl))
 	assert.NotNil(t, v2)
 }
+
+func TestGetProtocol(t *testing.T) {
+	singleton := GetProtocol()
+	assert.True(t, singleton == GetProtocol())
+}

--- a/registry/zookeeper/listener.go
+++ b/registry/zookeeper/listener.go
@@ -109,7 +109,13 @@ func (l *RegistryConfigurationListener) Next() (*registry.ServiceEvent, error) {
 	}
 }
 func (l *RegistryConfigurationListener) Close() {
-	l.client.Close()
+	if l.registry.IsAvailable() {
+		/**
+		 * if the registry is not available, it means that the registry has been destroy
+		 * so we don't need to call Done(), or it will cause the negative count panic for registry.wg
+		 */
+		l.registry.wg.Done()
+	}
 }
 
 func (l *RegistryConfigurationListener) valid() bool {

--- a/registry/zookeeper/listener.go
+++ b/registry/zookeeper/listener.go
@@ -109,7 +109,7 @@ func (l *RegistryConfigurationListener) Next() (*registry.ServiceEvent, error) {
 	}
 }
 func (l *RegistryConfigurationListener) Close() {
-	l.registry.wg.Done()
+	l.client.Close()
 }
 
 func (l *RegistryConfigurationListener) valid() bool {

